cpp/src/reductions/segmented_reductions.cpp

```diff
@@ -0,0 +1,103 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/reduction_functions.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf {
+namespace detail {
+struct segmented_reduce_dispatch_functor {
+  column_view const& col;
+  device_span<size_type const> offsets;
+  data_type output_dtype;
+  null_policy null_handling;
+  rmm::mr::device_memory_resource* mr;
+  rmm::cuda_stream_view stream;
+
+  segmented_reduce_dispatch_functor(column_view const& segmented_values,
+                                    device_span<size_type const> offsets,
+                                    data_type output_dtype,
+                                    null_policy null_handling,
+                                    rmm::cuda_stream_view stream,
+                                    rmm::mr::device_memory_resource* mr)
+    : col(segmented_values),
+      offsets(offsets),
+      output_dtype(output_dtype),
+      null_handling(null_handling),
+      mr(mr),
+      stream(stream)
+  {
+  }
+
+  template <aggregation::Kind k>
+  std::unique_ptr<column> operator()()
+  {
+    switch (k) {
+      case aggregation::SUM:
+        return reduction::segmented_sum(col, offsets, output_dtype, null_handling, stream, mr);
+      case aggregation::PRODUCT:
+        return reduction::segmented_product(col, offsets, output_dtype, null_handling, stream, mr);
+      case aggregation::MIN:
+        return reduction::segmented_min(col, offsets, output_dtype, null_handling, stream, mr);
+      case aggregation::MAX:
+        return reduction::segmented_max(col, offsets, output_dtype, null_handling, stream, mr);
+      case aggregation::ANY:
+        return reduction::segmented_any(col, offsets, output_dtype, null_handling, stream, mr);
+      case aggregation::ALL:
+        return reduction::segmented_all(col, offsets, output_dtype, null_handling, stream, mr);
+      default: CUDF_FAIL("Unsupported aggregation type.");
+    }
+  }
+};
+
+std::unique_ptr<column> segmented_reduce(column_view const& segmented_values,
+                                         device_span<size_type const> offsets,
+                                         aggregation const& agg,
+                                         data_type output_dtype,
+                                         null_policy null_handling,
+                                         rmm::cuda_stream_view stream,
+                                         rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(offsets.size() > 0, "`offsets` vector should have at least 1 element.");
```

Comment:
Since `offsets` are indexed as N+1 segments, a single element in `offsets` means there are zero segments. Does that mean the result is empty? If so, does the value of the single offset matter? Must it be `0` or `<= segmented_values.size()` or is there any bound on it at all?

Reply:
> a single element in offsets means there are zero segments. Does that mean the result is empty?

That's correct. That single value must be 0, however we don't check it because of https://github.com/rapidsai/cudf/issues/5505. Out of bound behaviors are undefined. Like `gather` if `DONT_CHECK` flag is specified. Part of this was copied from `segmented_shift.cu`.

I think we can start putting together a specification for `offsets` derived from [arrow columnar format](https://arrow.apache.org/docs/format/Columnar.html#variable-size-binary-layout). Thinking about situations of below:
- `length==0` case
- Can `offsets` only cover part of the data column?
- What's the behavior of out of bound indices?

Reply:
I mentions undefined behavior of out of bound indices in docstring.